### ISSUE TITLE
util::Result has confusing interface for std::*_ptr T

### ISSUE THREAD

**Issue by luke-jr** [bitcoin/bitcoin#26004]
`util::Result<std::unique_ptr<...>>` treats a nullptr as a `true` value. This has caused at least one bug so far.

Can we easily forbid passing nullptr in this way? Is there a better/safer solution?

**Comment by fanquake** [comment]
cc @ryanofsky

**Comment by luke-jr** [comment]
(For the mentioned bug, see https://github.com/bitcoin-core/gui/issues/661 and fix in #26005)

**Comment by MarcoFalke** [comment]
See also #24423

**Comment by ryanofsky** [comment]
> See also #24423

Forgot about that discussion. This does seem like a genuine use-case for `not_null`.

Another idea for the list:

5. Could replace `Result<unique_ptr<Wallet>>` with `ResultPtr<Wallet>` where a new `ResultPtr` class basically acts like `unique_ptr` except it has a `util::Error` constructor so it can hold error information, and is always set to an error or non-null pointer, never a plain null pointer. `ResultPtr<Wallet>` would be semantically equivalent to `Result<not_null<unique_ptr<Wallet>>>` except it would have nicer syntax and not require double dereferencing to call wallet methods.

**Comment by luke-jr** [comment]
The issue with `not_null` seems to be that in some cases, it might not detect it until runtime, and then throw an exception - which would have been a bug in this case (https://github.com/bitcoin-core/gui/issues/661) too.

Can we maybe throw an exception in debug builds, but in production builds simply treat nullptr assignments as clearing?

**Comment by luke-jr** [comment]
>5. Could replace Result<unique_ptr<Wallet>> with ResultPtr<Wallet> where a new ResultPtr class basically acts like unique_ptr except it has a util::Error constructor so it can hold error information, and is always set to an error or non-null pointer, never a plain null pointer. ResultPtr<Wallet> would be semantically equivalent to Result<not_null<unique_ptr<Wallet>>> except it would have nicer syntax and not require double dereferencing to call wallet methods.

Would this impede having a constant `Wallet` object per wallet?

**Comment by ryanofsky** [comment]
> The issue with `not_null` seems to be that in some cases, it might not detect it until runtime, and then throw an exception - which would have been a bug in this case ([bitcoin-core/gui#661](https://github.com/bitcoin-core/gui/issues/661)) too.

Yes, we would want to use a strict not_null implementation that makes it a compile error to try to assign a nullable pointer to a non-nullable one. Strict non_null pointers can only be initialized with special constructors, and if you call those constructors correctly, the compiler can guarantee the pointers are safe to dereference after that. The benefit of this is that it is easier to verify not_null pointers are constructed properly than it is to verify nullable pointers are safe to dereference every place they are  dereferenced.

> Can we maybe throw an exception in debug builds, but in production builds simply treat nullptr assignments as clearing?

It would be an implementation detail of the not_null class, but it could provide a constructor that only throws in debug builds. 

> Would this impede having a constant `Wallet` object per wallet?

I'm not even sure we have this currently. `interfaces::Wallet` objects are disposable wrappers around a more permanent `wallet::CWallet` object, but I think as long as pointers can be moved from, then object lifetimes / constness should be unaffected by things proposed here.

**Comment by luke-jr** [comment]
>I'm not even sure we have this currently. interfaces::Wallet objects are disposable wrappers around a more permanent wallet::CWallet object, but I think as long as pointers can be moved from, then object lifetimes / constness should be unaffected by things proposed here.

We don't, but I think we should. It would be nice to put the abstraction interface between all accesses - trying to make CWallet itself abstract has proven impractical (trying to add support for a Lightning wallet).

**Comment by willcl-ark** [comment]
Would it be OK to close this and keep tracking any `non_null` implementation work in #24423?

**Comment by maflcko** [comment]
> > See also [#24423](https://github.com/bitcoin/bitcoin/issues/24423)
> 
> Forgot about that discussion. This does seem like a genuine use-case for `not_null`.

https://github.com/bitcoin/bitcoin/pull/34844 implements this. It will look something like:

```diff
diff --git a/src/addrdb.cpp b/src/addrdb.cpp
index 5db632c7e1..c94438af99 100644
--- a/src/addrdb.cpp
+++ b/src/addrdb.cpp
@@ -195,3 +195,3 @@ void ReadFromStream(AddrMan& addr, DataStream& ssPeers)
 
-util::Result<std::unique_ptr<AddrMan>> LoadAddrman(const NetGroupManager& netgroupman, const ArgsManager& args)
+util::Result<util::NotNull<std::unique_ptr<AddrMan>>> LoadAddrman(const NetGroupManager& netgroupman, const ArgsManager& args)
 {
@@ -200,3 +200,3 @@ util::Result<std::unique_ptr<AddrMan>> LoadAddrman(const NetGroupManager& netgro
 
-    auto addrman{std::make_unique<AddrMan>(netgroupman, deterministic, /*consistency_check_ratio=*/check_addrman)};
+    auto addrman{util::NotNull{std::make_unique<AddrMan>(netgroupman, deterministic, /*consistency_check_ratio=*/check_addrman)}};
 
@@ -209,3 +209,3 @@ util::Result<std::unique_ptr<AddrMan>> LoadAddrman(const NetGroupManager& netgro
         // Addrman can be in an inconsistent state after failure, reset it
-        addrman = std::make_unique<AddrMan>(netgroupman, deterministic, /*consistency_check_ratio=*/check_addrman);
+        addrman = util::NotNull{std::make_unique<AddrMan>(netgroupman, deterministic, /*consistency_check_ratio=*/check_addrman)};
         LogInfo("Creating peers.dat because the file was not found (%s)", fs::quoted(fs::PathToString(path_addr)));
@@ -217,3 +217,3 @@ util::Result<std::unique_ptr<AddrMan>> LoadAddrman(const NetGroupManager& netgro
         // Addrman can be in an inconsistent state after failure, reset it
-        addrman = std::make_unique<AddrMan>(netgroupman, deterministic, /*consistency_check_ratio=*/check_addrman);
+        addrman = util::NotNull{std::make_unique<AddrMan>(netgroupman, deterministic, /*consistency_check_ratio=*/check_addrman)};
         LogWarning("Creating new peers.dat because the file version was not compatible (%s). Original backed up to peers.dat.bak", fs::quoted(fs::PathToString(path_addr)));
diff --git a/src/addrdb.h b/src/addrdb.h
index 44e51ce8be..ffbc70e76e 100644
--- a/src/addrdb.h
+++ b/src/addrdb.h
@@ -11,2 +11,3 @@
 #include <util/result.h>
+#include <util/pointers.h>
 
@@ -51,3 +52,3 @@ public:
 /** Returns an error string on failure */
-util::Result<std::unique_ptr<AddrMan>> LoadAddrman(const NetGroupManager& netgroupman, const ArgsManager& args);
+util::Result<util::NotNull<std::unique_ptr<AddrMan>>> LoadAddrman(const NetGroupManager& netgroupman, const ArgsManager& args);
 
```

Where returning a Result holding a nullptr (to addrman) is not possible.

**Comment by maflcko** [comment]
> The issue with `not_null` seems to be that in some cases, it might not detect it until runtime, and then throw an exception - which would have been a bug in this case ([bitcoin-core/gui#661](https://github.com/bitcoin-core/gui/issues/661)) too.

For reference in https://github.com/bitcoin/bitcoin/pull/34844, the NotNull constructor is explicit , so `util::Result<std::unique_ptr<Wallet>> restoreWallet` would turn into `util::Result<util::NotNull<std::unique_ptr<Wallet>>> restoreWallet` and require to type out NotNull. If that is violated, it will crash on construction, as opposed to on the nullptr deref later on. 

If an unchecked NotNull construction still goes through review, it seems no different than any other unchecked nullptr deref?

Also, I presume the ternary operator wouldn't work with `util::Result<util::NotNull<std::unique_ptr<Wallet>>>`, so the bugfix from 9f650062fc4350d8c50ac8890cddd4d99f20f895 could simply be enforced with a static_assert inside util::Result to reject all pointers?
